internal/configs/resource.go

```diff
@@ -696,39 +705,171 @@ func NewProviderConfigRefMapping(p *Provider) *ProviderConfigRefMapping {
 	return m
 }
 
-// TODO/Oleksandr: review the calls and provide a proper mapping
-func NewProviderConfigMappingFromRef(ref *ProviderConfigRef) *ProviderConfigRefMapping {
-	m := &ProviderConfigRefMapping{
-		Name:         ref.Name,
-		NameRange:    ref.NameRange,
-		providerType: ref.providerType,
+// TODO/Oleksandr: review calls and provide a proper alias
+func (m *ProviderConfigRefMapping) GetNoKeyAlias() string {
+	return m.Aliases[addrs.NoKey]
+}
+
+func (m *ProviderConfigRefMapping) decodeStaticAlias(eval *StaticEvaluator, forEach hcl.Expression) hcl.Diagnostics {
+	if m.Alias == nil {
+		return nil
 	}
 
-	if ref.Alias == "" {
-		return m
+	var diags hcl.Diagnostics
+
+	// There is no for each so we just put alias under NoKey.
+	if forEach == nil {
+		// TODO/Oleksandr: use static eval for alias expr?
+		aliasVal, aliasDiags := m.Alias.Value(nil)
+		diags = diags.Extend(aliasDiags)
+		if aliasDiags.HasErrors() {
+			return diags
+		}
+
+		if !aliasVal.Type().Equals(cty.String) {
+			// TODO/Oleksandr: add friendly error
+			return diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid alias",
+				Detail:   "",
+				Subject:  m.Alias.Range().Ptr(),
+			})
+		}
+
+		m.Aliases = map[addrs.InstanceKey]string{
+			addrs.NoKey: aliasVal.AsString(),
+		}
+
+		return nil
 	}
 
-	m.Aliases = map[addrs.InstanceKey]string{
-		addrs.NoKey: ref.Alias,
+	// TODO/Oleksandr: Is static identifier valid if we evaluate for_each for alias?
+	instances, instancesDiags := eval.Evaluate(forEach, StaticIdentifier{
+		Module:    eval.call.addr,
+		Subject:   "alias",
+		DeclRange: m.Alias.Range(),
+	})
+	diags = diags.Extend(instancesDiags)
+	if instancesDiags.HasErrors() {
```

Comment:
This piece is a bit interesting, if we can't evaluate for_each here, we are not allowed to use each.value in the providers expression.  However, that does not necessarily mean there's a config issue.  I glossed over this scenario in my prototype.

Ex:
```hcl
module "foo" {
  for_each = data.some.stuff # Static evaluation not allowed
  providers {
    random = random["foo"]
    other = random[local.other] # Similar question to L722 above
  }
}
```

An alternative approach would to use expr dependencies to see if for_each is needed to evaluate.

Pseudo:
```
refs = lang.ReferencesInExpr(aliasExpr)
if for_each in refs {
  try eval for_each
  if fail eval, return diags
  else, alias mapping logic
  return mapping
}
return simple alias
```

Reply:
Though that does depend on the question of if we want to allow raw strings or locals in the expression

Reply:
That's an excellent point, thank you!

We discussed it with @Evi1Pumpkin and decided it makes sense to have {NoKey: "alias"} mapping for cases when there is no instance references in the alias expression.

I also added `HasAlias` and `HasInstanceRefsInAlias` shortcuts to easily check if we need to iterate over all the instance keys on the graph part.

Updates are in [c661554](https://github.com/opentofu/opentofu/pull/1892/commits/c661554a4171c400383e43537e0982d810304ef1)